### PR TITLE
Encapsulate REST code in DynatraceClient

### PR DESCRIPTION
**Pull request by didiladi**
This commit introduces a new client (`DynatraceClient`) for performing
basic CRUD operations on any Dynatrace API supported by monaco.

It encapsulates the configuration-specific inconsistencies of certain
APIs in one place to provide a common interface to work with. After
all: A user of DynatraceClient shouldn't care about the implementation
details of each individual Dynatrace API.

Its design is intentionally not dependant on the Config and Environment
interfaces included in monaco. This makes sure, that DynatraceClient
can be used as a base for future tooling, which relies on a
standardized way to access Dynatrace APIs.

The new client implementation opens the door for future improvements,
like concurrent web request execution and the implementation of
various caching strategies.

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -87,6 +89,10 @@ func newApi(id string, apiPath string) Api {
 }
 
 func NewApi(id string, apiPath string) Api {
+
+	// TODO log warning if the user tries to create an API with a id not present in map above
```
Comment (UnseenWizzard) [inline pkg/api/api.go:93]: That is currently only possible if changing the code though, isn't it?

File: cmd/monaco/main.go [hunk 2]
```diff
@@ -351,10 +353,22 @@ func deleteConfigs(apis map[string]api.Api, environments map[string]environment.
 		for name, environment := range environments {
 			util.Log.Info("Deleting %d configs for environment %s...", len(configs), name)
 
+			apiToken, err := environment.GetToken()
+			if err != nil {
+				return err
+			}
+			client := rest.NewDynatraceClient(environment.GetEnvironmentUrl(), apiToken)
```
Comment (UnseenWizzard) [inline cmd/monaco/main.go:360]: Is there a reason to create one client per environment for execute and another for delete? 

Maybe we could create the client per env once somewhere early in the execution and reuse them - e.g. keep a map of env->client that can be accessed in both execute and delete?

**Comment by UnseenWizzard** [comment]
Cool fast way to move to fewer client objects and a simpler rest call wrapper to be reused!

File: pkg/api/api.go (hunk 1 above)
Comment (didiladi) [reply pkg/api/api.go:93]: I'm afraid, no. Any user can import a package (pkg/` folder) in an external application and use it. I guess we should really start with validating the arguments passed to our public API. That's not just here the case, but everywhere across `monaco`.

File: cmd/monaco/main.go (hunk 2 above)
Comment (didiladi) [reply cmd/monaco/main.go:360]: Good idea, I already thought of that. However, I wanted to keep it as simple as possible for now. Later, we can do a lot of improvements like parallel execution, caching of the initial GET request results, ... :smile:

File: pkg/api/api.go (hunk 1 above)
Comment (UnseenWizzard) [reply pkg/api/api.go:93]: ah, good point :+1:

**Comment by UnseenWizzard** [comment]
In how far is this still work in progress @didiladi ?

**Comment by didiladi** [comment]
@UnseenWizzard this is ready to be reviewed. Its currently in a state I'm comftable with. There are some follow-up actions to this - I don't want to make the review even more difficult:

* Properly handle the error in [request.go](https://github.com/dynatrace-oss/dynatrace-monitoring-as-code/pull/74/files#diff-76bc1ca3c519d2fed86bc08c4d9d9722951fc57cde145898dc447d8dc442e975R79)
* Paralled request execution ()
* Caching of the initial GET request - we currently perform the GET for every config, no matter if there was already one applied for the same API.
* To create just one client per environment (as @UnseenWizzard wrote [here](https://github.com/dynatrace-oss/dynatrace-monitoring-as-code/pull/74#discussion_r533957137))

If you find the time, I'd be very happy with a review @UnseenWizzard, @patrickpichler, @CruzanCaramele, @tatomir146 :small_airplane:

**Comment by didiladi** [comment]
Note that this PR breaks the public API of the `request` package. However, after merging this PR, we will have a stable interface which can be used by various other tools.

**Review by UnseenWizzard — DISMISSED** [review]

**Review by patrickpichler — APPROVED** [review]

**Review by patrickpichler — COMMENTED** [review]
I just noticed the integration tests are failing. 

Other than that, the PR looks good to me :+1:

**Comment by didiladi** [comment]
> Integration tests are failing

Yes, I'm on it. Doesn't seem to come from my changes though.

**Comment by didiladi** [comment]
It was not related to my changes:

```
Response was: {"error":{"code":429,"message":"Rate limit of 200 requests/minute exceeded! Reset time: 2020-12-17 16:37:47 GMT (1608223067402)"}}, responsible config: test-resources/integration-multi-environment/cinema-infrastructure/auto-tag/application-tagging.json
```

I just re-triggered the workflow - it worked now. I'll work on finishing the rate limit strategy (#89) next.

**Review by patrickpichler — APPROVED** [review]

**Comment by patrickpichler** [comment]
Wouldn't a rebase be better than a merge? 
That way we would have a clean linear history

**Comment by didiladi** [comment]
> Wouldn't a rebase be better than a merge?
> That way we would have a clean linear history

That's why we activated the "Squash Rebase" merge strategy :1st_place_medal:

**Comment by patrickpichler** [comment]
That is true ^^

**Pull request merged**